### PR TITLE
Add `DistributedTopology2D`

### LINKED ISSUE #323: Distributed topology objects
- divide domain into n chunks, would contain ClimaComm context

Consider the 1D case:
```
function DistributedIntervalTopology(mesh::IntervalMesh, commctx::ClimaComms.Context)
  # how many procs?
  # what pid am i?
  # do linear partition, figure out indices, neighbors
  ...
end
```

Can we just use `SpectralElementSpace1D`, or do we need a distributed version?
 

Need an interface to say:
- what are the element indices that need to get send to each process (stuff to copy to send buffer)
- what are the indices of the ghost elements belonging to each process (stuff to copy from receive buffer)

### PR DESCRIPTION
**Pull request by kpamnany**
Closes #323.

- [X] Integrates `ClimaComms.jl`
- [X] Adds `DistributedTopology2D` and unit tests
- [X] Updates `SpectralElementSpace2D` to support a distributed topology
- [X] Adds a `dss_2d!` for a distributed topology and unit tests

Some comments:
- Move/rename `Spaces.setup_comms` to something sensible
- Multiple comments are tagged with `XXX` for reviewer consideration
- DSS performance can be improved by changing `ghost_faces` and `ghost_vertices`  to store useful indices (such as the neighbor index)
- Names (of files, types and functions) need to be reconsidered (`internal_vertices` or `interior_vertices`? `local_elems` or `real_elems`? etc.)

This is squashed and tests pass locally so it's ready to merge once `ClimaComms` is [registered](https://github.com/JuliaRegistries/General/pull/52941).

Cc: @sriharshakandala

### CODE REVIEW

File: test/Topologies/distributed.jl [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    MPI.mpiexec() do cmd
+        Base.run(
+            `$cmd $oversubscribe -np $ntasks $(Base.julia_cmd()) --startup-file=no --project=$(Base.active_project()) $file`;
+            wait = true,
+        )
+        true
+    end
+end
+
+runmpi(joinpath(@__DIR__, "dtopo.jl"), ntasks = 4)
```
Comment (simonbyrne) [inline test/Topologies/distributed.jl:20]: It would probably be helpful if we could mock the ClimaComms context so that we can run the tests in the same process.

Comment (kpamnany) [reply test/Topologies/distributed.jl:20]: See https://github.com/CliMA/ClimaComms.jl/issues/7.

File: examples/bickleyjet/bickleyjet_cg_invariant_hypervisc.jl [hunk 2]
```diff
@@ -3,6 +3,14 @@ using LinearAlgebra
...
     Domains, Fields, Geometry, Meshes, Operators, Spaces, Topologies
 
+#=
+# set up communications for distributed topology
+using ClimaComms
+using ClimaCommsMPI
+const Context = ClimaCommsMPI.MPICommsContext
+const pid, nprocs = ClimaComms.init(Context)
+=#
+
```
Comment (simonbyrne) [inline examples/bickleyjet/bickleyjet_cg_invariant_hypervisc.jl:13]: Lets revert the changes to this file for now.

File: examples/Manifest.toml [hunk 3]
```diff
@@ -100,20 +100,37 @@ git-tree-sha1 = "f641eb0a4f00c343bbc32346e1217b86f3ce9dad"
...
 
+[[deps.ClimaComms]]
+deps = ["CUDA", "KernelAbstractions", "StaticArrays"]
+path = "../../ClimaComms.jl"
+uuid = "3a4d1b5c-c61d-41fd-a00a-5873ba7a1b0d"
+version = "0.1.0"
+
+[[deps.ClimaCommsMPI]]
+deps = ["ClimaComms", "KernelAbstractions", "MPI"]
+path = "../../ClimaComms.jl/backends/ClimaCommsMPI"
```
Comment (simonbyrne) [inline examples/Manifest.toml:128]: Need to use the registered version (once https://github.com/JuliaRegistries/General/pull/53183 is merged)

**Review by simonbyrne — APPROVED** [review]

File: Project.toml [hunk 4]
```diff
@@ -7,6 +7,8 @@ version = "0.6.0"
 Adapt = "79e6a3ab-5dfb-504d-930d-738a2a938a0e"
 BlockArrays = "8e7c35d0-a365-5155-bbbb-fb81a777f24e"
 CUDA = "052768ef-5323-5732-b1bb-66c8b64840ba"
+ClimaComms = "3a4d1b5c-c61d-41fd-a00a-5873ba7a1b0d"
+ClimaCommsMPI = "5f86816e-8b66-43b2-912e-75384f99de49"
```
Comment (simonbyrne) [inline Project.toml:11]: Can remove ClimaCommsMPI from the main Project.toml

File: test/Spaces/distributed/ddss2.jl [hunk 5]
```diff
@@ -0,0 +1,97 @@
...
+            9.5,
+            14.0,
+            18.5,
+            9.5,
+            14.0,
+            18.5,
+            9.5,
+            14.0,
+            18.5,
+        ]
```
Comment (charleskawczynski) [inline test/Spaces/distributed/ddss2.jl:52]: Can we turn off the formatter for this? (and below)
